### PR TITLE
all: Refactor the protocol/model interface a bit (ref #8981)

### PR DESCRIPTION
**Pull request by calmh**
These are refactoring changes made in the multiple-connection PR (#8918), broken out to reduce the noise there. Hence reviewing it requires some suspension of disbelief in terms of whether these changes are necessary -- they are, over there, but not necessarily by themselves over here... :) The changes are:

- In `lib/protocol`, `func (Connection) ID() DeviceID` is now called `func (Connection) DeviceID() DeviceID()` to disambiguate it from future other kind of IDs (the connection ID).
- Also in `lib/protocol`, all the callbacks used to take the device ID as the first parameter to explain where they came from. These now get the actual Connection instead, which is more specific and enables things like replying on the same connection. There's some juggling to accomplish this, because the actual protocol code that does the callbacks (`rawConnection`) doesn't have a clue about a "connection", it just has a reader and a writer for the bytes. Thus the callback interface is split up -- inside the protocol, it doesn't take a connection or device ID as parameter (`contextLessModel`), it is assumed that whoever created the connection knows what the callbacks refer to. Then there is a small wrapper (`connectionWrappingModel`) struct that does precisely this: intercepts the callbacks, adds the connection info, and then calls them again upwards. 
- Changes in model to the new signatures.
- Changes in all the tests for the new signatures.
- Minor tweaks to some tests while I was anywhere there... (parallelization, listening on 127.0.0.1 instead of 0.0.0.0 when possible.)
- Harmonized log lines from showing full device ID to just short ID in places where I saw it and were anyway touching stuff...

### CODE REVIEW

File: lib/protocol/protocol.go [hunk 1]
```diff
@@ -121,19 +121,34 @@ var (
 	errFileHasNoBlocks    = errors.New("file with empty block list")
 )
 
+type contextLessModel interface {
+	// An index was received from the peer device
```
Comment (imsodin) [inline lib/protocol/protocol.go:125]: Instead of repeating the comment, I'd add an interface level comment explaining that this is 1-to-1 `Model` without the initial `Connection` param.

File: lib/protocol/protocol.go [hunk 2]
```diff
@@ -1068,3 +1094,35 @@ func messageContext(msg message) (string, error) {
 		return "", errors.New("unknown or empty message")
 	}
 }
+
+// connectionWrappingModel takes the Model interface from the model package,
+// which expects the Connection as the first parameter in all methods, and
+// wraps it to conform to the protocol.Model interface.
```
Comment (imsodin) [inline lib/protocol/protocol.go:1100]:
```suggestion
// wraps it to conform to the protocol.contextLessModel interface.
```

File: lib/protocol/protocol.go [hunk 3]
```diff
@@ -938,9 +960,13 @@ func (c *rawConnection) internalClose(err error) {
 		}
 		c.awaitingMut.Unlock()
 
-		<-c.dispatcherLoopStopped
+		if !c.startTime.IsZero() {
+			// Wait for the dispatcher loop to exit, if it was started to
+			// begin with.
+			<-c.dispatcherLoopStopped
+		}
```
Comment (imsodin) [inline lib/protocol/protocol.go:967]: I guess that caused issues during testing/debugging? I'd put that into a separate commit/PR then, just so we don't lose this fix if for some reason this entire refactoring gets reverted.

**Review by imsodin — COMMENTED** [review]
Definitely very helpful to separate this large code churn from the large behaviour change - looks mostly good on a skim.

**Review by imsodin — APPROVED** [review]
Unrelated: A filesystem watcher test times out on macos - maybe truly a (flaky) hangup or possibly just need to extend the timeout as GHA macos FS might just be super slow.

**Comment by calmh** [comment]
Yeah it's super flaky, I have on my mental todo list to go through and unflake all the stuff that keeps randomly failing.

**Comment by tomasz1986** [comment]
The ref at https://github.com/syncthing/syncthing/issues/8981 seems like a typo.

**Comment by imsodin** [comment]
Yep, should have been #8918 - I just assumed it was without checking - bad reviewer, assuming is no good there :)
